### ISSUE TITLE
components that write to their own parameter properties

### ISSUE THREAD

**Issue by enetstudio** [dotnet/aspnetcore#21317]
I'm referring to [Don't create components that write to their own parameter properties](https://docs.microsoft.com/en-us/aspnet/core/blazor/components?view=aspnetcore-3.1#dont-create-components-that-write-to-their-own-parameter-properties), where we're told that  

> Parameters are overwritten under the following conditions:

> *   A child component's content is rendered with a RenderFragment.
> *   StateHasChanged is called in the parent component

My question is how do you account to the resetting of the Count parameter in a child component in the following code snippet:

     @page "/"
    
     <p>Count value in parent: @countOutput</p>

     <ChildComponent Count="10" PassBack="@Count"/>
   
    @code
    {
        private string countOutput;

        private void Count(int count)
        {
            countOutput = count.ToString();
        }
    }

ChildComponent.razor
-------------------------
    <p>Count in child component: @Count.ToString()</p>

    <button type="button" 
                  @onclick="IncrementParameter">Increment</button>

    @code {
        [Parameter]
        public int Count { get; set; }

        [Parameter]
        public EventCallback<int> PassBack { get; set; }

        private async Task IncrementParameter()
       {
            Count++;
            await Task.Delay(3000);
            if (PassBack.HasDelegate)
           {
               await PassBack.InvokeAsync(Count);
           }
       }
    }

Please note that the Count parameter is always reset to 10, and of course,  the  value passed back to the parent component and displayed is always 11. 

As you can see, no child content is involved here. What you guys teach in your docs is correct, but I see that there is more to it than said in your docs.

Many thanks...

**Comment by javiercn** [comment]
@enetstudio thanks for contacting us.

You override `OnParameterSetAsync` and reset the internal state for your component. You should not be doing Count++ but have a separate _count field to track the state, that's precisely why we have the guidance that you shouldn't write the parameters.

**Comment by enetstudio** [comment]
@javiercn, I thank you for your response.

> You should not be doing Count++ but have a separate _count field to track the state, that's precisely why we have the guidance that you shouldn't write the parameters

Yes, I understand that.  And even before I've discovered this section in the docs I knew that I must not manipulate the parameter itself, and should assign it to a local variable. However, this issue is written because the docs speak of two factors or conditions: Child content in child component and StateHasChanged call in parent component. In my case, there is no child content; only implicit call to StateHasChanged in parent component, and manipulating the parameter in child component, and yet the same behavior is  exhibited. My personal interpretation is that it has nothing to do with child content, and the issue stems from the fact that the parameter is manipulated, and as long as I don't call the StateHasChanged method in the parent component, the parameter keeps incrementing, but once I call the StateHasChanged, the parameter is reset...

Many thanks...

**Comment by javiercn** [comment]
@enetstudio yep, that is correct. It has nothing to do with ChildContent, it's because the parameter is being manipulated.

Closing since the behavior is expected and the question has been answered.

**Comment by enetstudio** [comment]
Many thanks to you...
